lib/logger.js

```diff
@@ -0,0 +1,589 @@
+'use strict';
+const {
+  DateNow,
+  JSONStringify,
+  ObjectAssign,
+  ObjectHasOwn,
+  ObjectKeys,
+  SafeSet,
+} = primordials;
+
+const { isNativeError } = require('internal/util/types');
+
+
+const {
+  codes: {
+    ERR_INVALID_ARG_TYPE,
+    ERR_INVALID_ARG_VALUE,
+    ERR_METHOD_NOT_IMPLEMENTED,
+  },
+} = require('internal/errors');
+
+const {
+  validateObject,
+  validateString,
+  validateOneOf,
+  validateFunction,
+} = require('internal/validators');
+
+const Utf8Stream = require('internal/streams/fast-utf8-stream');
+const diagnosticsChannel = require('diagnostics_channel');
+const { kEmptyObject } = require('internal/util');
+const stdSerializers = require('internal/logger/serializers');
+
+// Create channels for each log level
+const channels = {
+  trace: diagnosticsChannel.channel('log:trace'),
+  debug: diagnosticsChannel.channel('log:debug'),
+  info: diagnosticsChannel.channel('log:info'),
+  warn: diagnosticsChannel.channel('log:warn'),
+  error: diagnosticsChannel.channel('log:error'),
+  fatal: diagnosticsChannel.channel('log:fatal'),
+};
+
+/*
+* RFC5424 numerical ordering + log4j interface
+* @link https://www.rfc-editor.org/rfc/rfc5424.html
+*/
+const LEVELS = {
+  trace: 10,
+  debug: 20,
+  info: 30,
+  warn: 40,
+  error: 50,
+  fatal: 60,
+};
+
+const LEVEL_NAMES = ObjectKeys(LEVELS);
+
+// Noop function for disabled log levels
+function noop() {}
+
+/**
+ * Base consumer class for handling log records
+ * Consumers subscribe to diagnostics_channel events
+ */
+class LogConsumer {
+  #levelValue;
+
+  constructor(options = kEmptyObject) {
+    validateObject(options, 'options');
+    const { level = 'info' } = options;
+    validateString(level, 'options.level');
+
+    validateOneOf(level, 'options.level', LEVEL_NAMES);
+
+    this.#levelValue = LEVELS[level];
+  }
+
+  /**
+   * Check if a level would be consumed
+   * @param {string} level
+   * @returns {boolean}
+   */
+  enabled(level) {
+    if (!ObjectHasOwn(LEVELS, level)) {
+      return false;
+    }
+    return LEVELS[level] >= this.#levelValue;
+  }
+
+  /**
+   * Attach this consumer to log channels
+   */
+  attach() {
+    for (const level of LEVEL_NAMES) {
+      if (this.enabled(level)) {
+        channels[level].subscribe(this.#handleLog.bind(this, level));
+      }
+    }
+  }
+
+  /**
+   * Internal handler called by diagnostics_channel
+   * @param {string} level
+   * @param {object} record
+   * @private
+   */
+  #handleLog(level, record) {
+    this.handle(record);
+  }
+
+  /**
+   * Handle a log record (must be implemented by subclass)
+   * @param {object} record
+   */
+  handle(record) {
+    throw new ERR_METHOD_NOT_IMPLEMENTED('LogConsumer.handle()');
+  }
+}
+
+/**
+ * JSON consumer - outputs structured JSON logs
+ */
+class JSONConsumer extends LogConsumer {
+  #stream;
+  #fields;
+
+  constructor(options = kEmptyObject) {
```

Comment:
Can you add JSDoc to all methods with proper types? It's hard to follow up on the parameters and what options are supported etc.

Reply:
Thanks for all suggestions, I tryed edited @anonrig